### PR TITLE
feat(provider): new strategy for auth0

### LINKED ISSUE #244: Authentication using Auth0 strategy
**Is your feature request related to a problem? Please describe.**
Right now we dont have auth0 passport strategy 
**Describe the solution you'd like**
Authentication using auth0 strategy would be really helpful

### PR DESCRIPTION
**Pull request by yeshamavani**
passport based strategy

GH-244

## Description

Added a new passport based strategy for auth0 authentication.

Fixes #244

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Used this package in sourceloop based authentication service 
## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the style guide
- [x] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/strategies/passport/passport-auth0/auth0-strategy-factory-provider.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  StrategyOption,
+  StrategyOptionWithRequest,
+} from 'passport-auth0';
+import {Request} from 'express';
+
+export interface Auth0StrategyOptions extends StrategyOption {
+  passReqToCallback?: false | undefined;
+}
+/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+export type VerifyCallback = (err?: any, user?: any, info?: any) => void;
```
Comment (samarpanB) [inline src/strategies/passport/passport-auth0/auth0-strategy-factory-provider.ts:19]: Isnt this available in passport strategy typings

File: src/strategies/types/types.ts [hunk 2]
```diff
@@ -120,6 +121,16 @@ export namespace VerifyFunction {
       req?: Request,
     ): Promise<T | null>;
   }
+  export interface Auth0Fn<T = IAuthUser> extends GenericAuthFn<T> {
```
Comment (samarpanB) [inline src/strategies/types/types.ts:124]: Why this ?

File: src/strategies/user-auth-strategy.provider.ts [hunk 3]
```diff
@@ -241,6 +241,25 @@ export class AuthStrategyProvider implements Provider<Strategy | undefined> {
...
+    if (!auth0Factory) {
+      throw new Error(
+        `No factory found for ${Strategies.Passport.AUTH0_STRATEGY_FACTORY}`,
+      );
+    }
+
+    // Cast the factory output to `Strategy` type
+    return auth0Factory(
+      /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+      this.metadata.options as any, //NOSONAR
```
Comment (samarpanB) [inline src/strategies/user-auth-strategy.provider.ts:259]: why any ?

**Review by samarpanB — COMMENTED** [review]
pls check the comments

File: src/strategies/types/types.ts (hunk 2 above)
Comment (yeshamavani) [reply src/strategies/types/types.ts:130]: We create interface for each strategy based on its verify function type

File: src/strategies/passport/passport-auth0/auth0-strategy-factory-provider.ts (hunk 1 above)
Comment (yeshamavani) [reply src/strategies/passport/passport-auth0/auth0-strategy-factory-provider.ts]: no not explicitly

**Pull request merged**